src/backend/distributed/utils/node_metadata.c

```diff
@@ -145,6 +148,67 @@ master_initialize_node_metadata(PG_FUNCTION_ARGS)
 
 
 /*
+ * get_shard_id_for_distribution_column function takes a distributed table name and a
+ * distribution value then returns shard id of the shard which belongs to given table and
+ * contains given value. This function only works for hash distributed tables.
+ */
+Datum
+get_shard_id_for_distribution_column(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	Datum distributionValue = PG_GETARG_DATUM(1);
+
+	Var *distributionColumn = NULL;
+	char distributionMethod = 0;
+	Oid expectedElementType = InvalidOid;
+	Oid inputElementType = InvalidOid;
+	DistTableCacheEntry *cacheEntry = NULL;
+	int shardCount = 0;
+	ShardInterval **shardIntervalArray = NULL;
+	FmgrInfo *hashFunction = NULL;
+	FmgrInfo *compareFunction = NULL;
+	bool useBinarySearch = true;
+	ShardInterval *shardInterval = NULL;
+
+	if (!IsDistributedTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+						errmsg("relation is not distributed")));
+	}
+
+	distributionMethod = PartitionMethod(relationId);
+	if (distributionMethod != DISTRIBUTE_BY_HASH)
```

Comment:
We could also support DISTRIBUTE_BY_RANGE and (coming soon) DISTRIBUTE_BY_NONE. It could even be somewhat useful to make this work for DISTRIBUTE_BY_APPEND, though it would have to return an array probably.